### PR TITLE
Add more inplace arguments to quantization top level API

### PR DESCRIPTION
**Pull request by dzhulgakov**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26782 Add more inplace arguments to quantization top level API**

At least we should be consistent on top-level APIs and prepare/convert/etc.

Logic is inplace=False by default but top-level APIs take care of doing fewer copies.

Also renames always-inplace methods like add_observer to have underscore in the end.

One fix for MinMaxObserver was triggered by deepcopy surfacing that we were accidentally keeping autograd around

Differential Revision: [D17595956](https://our.internmc.facebook.com/intern/diff/D17595956)

### CODE REVIEW

File: torch/quantization/quantize.py [hunk 1]
```diff
@@ -176,20 +177,33 @@ def add_quant_dequant(module):
...
+    will be propagated.
 
     Args:
-        mod: input model
+        model: input model to be modified in-place
+        qconfig_dict: dictionary that maps from name or type of submodule to quantization
+            configuration, qconfig applies to all submodules of a given
+            module unless qconfig for the submodules are specified (when the
+            submodule already has qconfig attribute)
+        inplace: carry out model transformations in-place, the original model is destroyed
```
Comment (ZolotukhinM) [inline torch/quantization/quantize.py]: Nit: is "destroyed" the right word here? It might look to users that if I do `prepare(model)` then the `model` is destroyed and I cannot use it anymore.

**Review by ZolotukhinM — COMMENTED** [review]
Looks good to me, but I'll let other people to review it.

File: torch/quantization/quantize.py [hunk 2]
```diff
@@ -241,19 +255,22 @@ def quantize(model, run_fn, run_args, mapping=DEFAULT_MODULE_MAPPING):
...
     Return:
         Quantized model.
     """
 
-    model = copy.deepcopy(model)
+    if not inplace:
+        model = copy.deepcopy(model)
     model.eval()
-    prepare(model)
+    model = prepare(model, inplace=True)
```
Comment (jerryzh168) [inline torch/quantization/quantize.py]: why do we need to assign with `inplace=True`?

File: torch/quantization/quantize.py [hunk 3]
```diff
@@ -176,20 +177,33 @@ def add_quant_dequant(module):
...
+            submodule already has qconfig attribute)
+        inplace: carry out model transformations in-place, the original model is destroyed
     """
-    propagate_qconfig(model)
-    add_observer(model)
+    if not inplace:
+        model = copy.deepcopy(model)
+    propagate_qconfig_(model)
+    # sanity check common API misusage
+    if not any(hasattr(m, 'qconfig') and m.qconfig for m in model.modules()):
```
Comment (jerryzh168) [inline torch/quantization/quantize.py]: although this is a warning, but maybe we want to check all the modules in the module hierarchy to say there is no qconfig applied?

File: torch/quantization/quantize.py [hunk 4]
```diff
@@ -290,17 +308,19 @@ def quantize_dynamic(model, qconfig_dict=None, dtype=torch.qint8, mapping=DEFAUL
...
                 "Don't know how to quantize with default settings for {}. Provide full qconfig please".format(dtype))
 
-    model = copy.deepcopy(model)
+    if not inplace:
+        model = copy.deepcopy(model)
     model.eval()
-    propagate_qconfig(model, qconfig_dict)
-    convert(model, mapping)
+    propagate_qconfig_(model, qconfig_dict)
+    model = convert(model, mapping, inplace=True)
```
Comment (jerryzh168) [inline torch/quantization/quantize.py]: why do we need to return with inplace=True?

File: torch/quantization/quantize.py (hunk 2 above)
Comment (dzhulgakov) [reply torch/quantization/quantize.py]: just for consistency with potential graph mode

File: torch/quantization/quantize.py (hunk 3 above)
Comment (dzhulgakov) [reply torch/quantization/quantize.py]: that's what it does. .modules() returns all submodules recursively

File: test/test_quantization.py [hunk 5]
```diff
@@ -59,25 +59,33 @@ def checkQuantized(model):
...
+        base = SingleLayerLinearModel()
+        keys_before = set(list(base.state_dict().keys()))
+        model = quantize(base, test_only_eval_fn, self.calib_data)
+        checkQuantized(model)
+        keys_after = set(list(base.state_dict().keys()))
+        self.assertEqual(keys_before, keys_after)  # simple check that nothing changed
+
+        # in-place version
+        model = SingleLayerLinearModel()
+        quantize(model, test_only_eval_fn, self.calib_data, inplace=True)
```
Comment (jerryzh168) [inline test/test_quantization.py]: for`inplace=True` should we either always assign or not assign the result?

Comment (dzhulgakov) [reply test/test_quantization.py]: good catch, I might have missed a few places

**Comment by dzhulgakov** [comment]
Actually, I think I'll keep unittests with out assignment if inplace=True. It should be doable with graph mode too (by assigning the graph in-place) and APIs look much nicer without the need for extra assignemnt

**Comment by raghuramank100** [comment]
> Stack from [ghstack](https://github.com/ezyang/ghstack):
> 
> * **#26782 Add more inplace arguments to quantization top level API**
> 
> At least we should be consistent on top-level APIs and prepare/convert/etc.
> 
> Logic is inplace=False by default but top-level APIs take care of doing fewer copies.
> 
> Also renames always-inplace methods like add_observer to have underscore in the end.
> 
> One fix for MinMaxObserver was triggered by deepcopy surfacing that we were accidentally keeping autograd around

The minmaxobserver fix also showed up as part of QAT debug, but my diffs havent landed yet, all observers now have with torch.no_grad() or detach to prevent this.

**Review by raghuramank100 — APPROVED** [review]

File: test/test_quantized_nn_mods.py [hunk 6]
```diff
@@ -300,11 +300,11 @@ def test_linear_api(self, batch_size, in_features, out_features, use_bias, use_f
         # Test from_float.
         float_linear = torch.nn.Linear(in_features, out_features).float()
         float_linear.qconfig = torch.quantization.default_qconfig
-        torch.quantization.prepare(float_linear)
+        torch.quantization.prepare(float_linear, inplace=True)
         float_linear(X.float())
         # Sequential allows swapping using "convert".
         quantized_float_linear = torch.nn.Sequential(float_linear)
-        torch.quantization.convert(quantized_float_linear)
+        quantized_float_linear = torch.quantization.convert(quantized_float_linear, inplace=True)
```
Comment (jerryzh168) [inline test/test_quantized_nn_mods.py]: one missed assignment here

File: torch/quantization/quantize.py [hunk 7]
```diff
@@ -176,20 +177,33 @@ def add_quant_dequant(module):
...
+            configuration, qconfig applies to all submodules of a given
+            module unless qconfig for the submodules are specified (when the
+            submodule already has qconfig attribute)
+        inplace: carry out model transformations in-place, the original module is mutated
     """
-    propagate_qconfig(model)
-    add_observer(model)
+    if not inplace:
+        model = copy.deepcopy(model)
+    propagate_qconfig_(model)
```
Comment (masahi) [inline torch/quantization/quantize.py]: why qconfig_dict is not passed to propagate_qconfig_(...)? qconfig_dict is not used inside this function.
Is it intended?
